src/ResponseBuilder.cpp

```diff
@@ -0,0 +1,176 @@
+ * Else skip this step.
+ * Execute response body handler to get the response body.
+ * Build the response by appending the status line, headers and body.
+ * @param request HTTP request.
+ */
+void ResponseBuilder::buildResponse(const HTTPRequest& request)
+{
+	resetStream();
+
+	HTTPResponse response = initHTTPResponse(request);
+
+	if (response.status == StatusOK) {
+		TargetResourceHandler targetResourceHandler(m_activeServer->locations, request, response, m_fileSystemPolicy);
+		targetResourceHandler.execute();
+	}
+
+	ResponseBodyHandler responseBodyHandler(response, m_fileSystemPolicy);
+	responseBodyHandler.execute();
+
+	appendStatusLine(response);
+	appendHeaders(response);
+	m_response << response.body << "\r\n";
+}
+
+/**
+ * @brief Reset the stream to the beginning.
+ *
+ * If it is not the first time, the stream is reset to the beginning.
+ * This is done to clear the stream from the previous response.
+ */
+void ResponseBuilder::resetStream()
+{
+	if (!m_isFirstTime)
+		m_response.seekp(std::ios::beg);
+
+	m_isFirstTime = false;
+}
+
+/**
+ * @brief Init the HTTP Response object with data from the request.
+ *
+ * @param request HTTP request.
+ * @return HTTPResponse Constructed HTTP response.
+ */
+HTTPResponse ResponseBuilder::initHTTPResponse(const HTTPRequest& request)
+{
+	HTTPResponse response;
+
+	response.status = StatusOK;
+	response.method = request.method;
+	response.autoindex = false;
+
+	return response;
+}
+
+/**
+ * @brief Append status line to the response.
+ *
+ * The status line is appended in the following format:
+ * HTTP/1.1 <status code> <reason phrase>
+ * @param response HTTP response.
+ */
+void ResponseBuilder::appendStatusLine(const HTTPResponse& response)
+{
+	m_response << "HTTP/1.1 " << response.status << ' ' << utils::statusCodeToReasonPhrase(response.status) << "\r\n";
+}
+
+/**
+ * @brief Append headers to the response.
+ *
+ * The following headers are appended:
+ * Content-Type: MIME type of the target resource.
+ * Content-Length: Length of the response body.
+ * Server: TriHard.
+ * Date: Current date in GMT.
+ * Location: Target resource if status is StatusMovedPermanently.
+ * Delimiter.
+ * @param response HTTP response.
+ */
+void ResponseBuilder::appendHeaders(const HTTPResponse& response)
+{
+	// Content-Type
+	m_response << "Content-Type: " << getMIMEType(utils::getFileExtension(response.targetResource)) << "\r\n";
+	// Content-Length
+	m_response << "Content-Length: " << response.body.length() << "\r\n";
+	// Server
+	m_response << "Server: TriHard\r\n";
+	// Date
+	m_response << "Date: " << utils::getGMTString(time(0), "%a, %d %b %Y %H:%M:%S GMT") << "\r\n";
+	// Location
+	if (response.status == StatusMovedPermanently) {
+		m_response << "Location: " << response.targetResource << "\r\n";
+	}
+	// Delimiter
+	m_response << "\r\n";
+}
+
+/**
+ * @brief Init the MIME types map.
+ *
+ * @todo FIXME: This needs to get the MIME type of the found location.
```

Comment:
This is a temporary.

MIME (Multipurpose Internet Mail Extensions) types indicate the nature and format of a document, file, or assortment of bytes on the server:
- html pages have text/html
- css have text/css
- images can have image/jpeg or image/png

It stems from emails but the name seems to have stuck. 
For a longer explanation of MIME types see for example this artivle from Mozilla [LINK](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types)

They are standardized in IETF's [RFC 6838](https://datatracker.ietf.org/doc/html/rfc6838) 
But they can often be freely defined on the server. For example nginx has the types directive: [LINK](https://nginx.org/en/docs/http/ngx_http_core_module.html#types). This can be done on server and location level

Since the interface on how we would store the mime types is not defined yet, the response builder just has some standard mime types set. It just fetches them from its internal storage. Later on this would use the found location to fetch the correct type from there.

Reply:
Yes, I meant how do you imagine this location-specific assignment with this function?

Reply:
More specific question would help.

The two functions initMIMETypes and getMIMEType have the same @todo, since they are both dependent on further implementation details, i.e. how we store the MIME types.

Right now TargetResourceHandler saves reference of found location in HTTP Response. getMIMEType can fetch the map, vector, whatever of this location and search for type.

I imagine two possible implementations (right now):
1. we save in each location all MIME types (specific and default ones)
2. we only save MIME types we have explicitly defined and have defaults in ResponseBuilder

In case 1 we don't need this map, so I would delete. In case 2 initMIMETypes inits the map of types saved in ResponseBuilder. Maybe there are other possibilities.

Reply:
Best to talk about this in person to clear it up.